invokeai/app/services/image_record_storage.py

```diff
@@ -0,0 +1,353 @@
+        per_page: int = 10,
+    ) -> PaginatedResults[ImageRecord]:
+        """Gets a page of image records."""
+        pass
+
+    @abstractmethod
+    def delete(self, image_type: ImageType, image_name: str) -> None:
+        """Deletes an image record."""
+        pass
+
+    @abstractmethod
+    def save(
+        self,
+        image_name: str,
+        image_type: ImageType,
+        image_category: ImageCategory,
+        session_id: Optional[str],
+        node_id: Optional[str],
+        metadata: Optional[ImageMetadata],
+        created_at: str = datetime.datetime.utcnow().isoformat(),
+    ) -> None:
+        """Saves an image record."""
+        pass
+
+
+class SqliteImageRecordStorage(ImageRecordStorageBase):
+    _filename: str
+    _conn: sqlite3.Connection
+    _cursor: sqlite3.Cursor
+    _lock: threading.Lock
+
+    def __init__(self, filename: str) -> None:
+        super().__init__()
+
+        self._filename = filename
+        self._conn = sqlite3.connect(filename, check_same_thread=False)
+        # Enable row factory to get rows as dictionaries (must be done before making the cursor!)
+        self._conn.row_factory = sqlite3.Row
+        self._cursor = self._conn.cursor()
+        self._lock = threading.Lock()
+
+        try:
+            self._lock.acquire()
+            # Enable foreign keys
+            self._conn.execute("PRAGMA foreign_keys = ON;")
+            self._create_tables()
+            self._conn.commit()
+        finally:
+            self._lock.release()
+
+    def _create_tables(self) -> None:
+        """Creates the tables for the `images` database."""
+
+        # Create the `images` table.
+        self._cursor.execute(
+            f"""--sql
+            CREATE TABLE IF NOT EXISTS images (
+                id TEXT PRIMARY KEY,
+                image_type TEXT, -- non-nullable via foreign key constraint
+                image_category TEXT, -- non-nullable via foreign key constraint
+                session_id TEXT, -- nullable
+                node_id TEXT, -- nullable
+                metadata TEXT, -- nullable
+                created_at TEXT NOT NULL,
+                FOREIGN KEY(image_type) REFERENCES image_types(type_name),
+                FOREIGN KEY(image_category) REFERENCES image_categories(category_name)
+            );
+            """
+        )
+
+        # Create the `images` table indices.
+        self._cursor.execute(
+            """--sql
+            CREATE UNIQUE INDEX IF NOT EXISTS idx_images_id ON images(id);
+            """
+        )
+        self._cursor.execute(
+            """--sql
+            CREATE INDEX IF NOT EXISTS idx_images_image_type ON images(image_type);
+            """
+        )
+        self._cursor.execute(
+            """--sql
+            CREATE INDEX IF NOT EXISTS idx_images_image_category ON images(image_category);
+            """
+        )
+
+        # Create the tables for image-related enums
+        create_enum_table(
```

Comment:
This could also be a good alternative to avoid using an enum DB field: https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#using-python-enum-or-pep-586-literal-types-in-the-type-map

```
class Status(enum.Enum):
    PENDING = "pending"
    RECEIVED = "received"
    COMPLETED = "completed"


class SomeClass(Base):
    __tablename__ = "some_table"

    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[Status]
```

Reply:
If categories are added, it inserts another row on startup.

There's a less flexible version of enums using the `CHECK` syntax, which sets the valid values on table creation (and you have to do a migration to change it).

I don't mind demoting this to just a `TEXT` column and adding some comments to the enums that are used to populate the column - my first thought was to model it like an Enum.